### DISCUSSION TITLE
CDK v2 Tags assertion

### DISCUSSION THREAD

**Discussion by caevv** [aws/aws-cdk#21282] [Q&A]
How do you assert tags?

```ts
import { Template } from "aws-cdk-lib/assertions";

describe("Stack", () => {
  const app = new App();

  it("tags", () => {
    const stack = new Stack(app, "my-stack", {
      tags: {
        foo: "bar",
      },
    });
    new s3.Bucket(stack, "myS3");

    const template = Template.fromStack(stack);

    template.hasResource("AWS::S3::Bucket", {
      Tags: [
        {
          foo: "bar",
        },
      ],
    });
  });
});
```

`Error: Template has 1 resources with type AWS::S3::Bucket, but none match as expected.`

**Comment by pflorek** [comment]
Almost the same way. `template.hasResourceProperties("AWS::S3::Bucket"` is shorthand for `template.hasResource("AWS::S3::Bucket", { Properties:` See https://docs.aws.amazon.com/cdk/api/v1/docs/assertions-readme.html#resource-matching--retrieval

The `Tags` of the AWS::S3::Bucket are in a different format https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html

```typescript
template.hasResourceProperties("AWS::S3::Bucket", {
      Tags: [
        {
          Key: "foo",
          Value: "bar",
        },
      ],
    });
```

See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html

You may use also `Tags.of(scope).addTag("foo", "bar");` anywhere to tag a specific resource or all children in the construct tree https://docs.aws.amazon.com/cdk/v2/guide/tagging.html
